### PR TITLE
Add support for Horovod (tensorflow)

### PR DESCRIPTION
**Pull request by pescap**
Work in progress. Add the hvd support, to use multiple GPUs. Sounds to be working properly (tested with 1 up to 8 GPUs) and is non-intrusive! Later, we could add more the hvd-accelerated `tf.compat.v1` and `pytorch` backends.

To do:

- [ ] Avoid multiple imports of `import horovod.tensorflow as hvd`. I have to understand how to import horovod only once. I wonder where I shall put this import. I do not want the code to depend on `hvd`, if one does not want to use multiple GPU acceleration. 
- [ ] Define properly the collocation points. The idea would be to generate independent samples for each GPU. This is not done properly so far. With the current code, I am sending the same points to all the GPUs, and I add a resampling process `resampler = dde.callbacks.PDEResidualResampler(period= epochs // hvd.size())`. To my opinion, the most simple option would be to set properly the mini_batch option (see #320 )... Do you agree? 

Tested with the 2D Helmhlotz case (notice that I only added the resampler and `dde.config.set_hvd()`):

```
"""Backend supported: tensorflow.compat.v1, tensorflow, pytorch"""
import deepxde as dde
import numpy as np

dde.config.set_hvd()

# Set seed on 11
dde.config.set_random_seed(11)
# General parameters
n = 2
precision_train = 10
precision_test = 30
hard_constraint = True
weights = 100  # if hard_constraint == False
epochs = 5000
parameters = [1e-3, 3, 150, "sin"]

# Define sine function
if dde.backend.backend_name == "pytorch":
    sin = dde.backend.pytorch.sin
else:
    from deepxde.backend import tf

    sin = tf.sin

learning_rate, num_dense_layers, num_dense_nodes, activation = parameters


def pde(x, y):
    dy_xx = dde.grad.hessian(y, x, i=0, j=0)
    dy_yy = dde.grad.hessian(y, x, i=1, j=1)

    f = k0 ** 2 * sin(k0 * x[:, 0:1]) * sin(k0 * x[:, 1:2])
    return -dy_xx - dy_yy - k0 ** 2 * y - f


def func(x):
    return np.sin(k0 * x[:, 0:1]) * np.sin(k0 * x[:, 1:2])


def transform(x, y):
    res = x[:, 0:1] * (1 - x[:, 0:1]) * x[:, 1:2] * (1 - x[:, 1:2])
    return res * y


def boundary(_, on_boundary):
    return on_boundary


geom = dde.geometry.Rectangle([0, 0], [1, 1])
k0 = 2 * np.pi * n
wave_len = 1 / n

hx_train = wave_len / precision_train
nx_train = int(1 / hx_train)

hx_test = wave_len / precision_test
nx_test = int(1 / hx_test)

if hard_constraint == True:
    bc = []
else:
    bc = dde.icbc.DirichletBC(geom, lambda x: 0, boundary)


data = dde.data.PDE(
    geom,
    pde,
    bc,
    num_domain=nx_train ** 2,
    num_boundary=4 * nx_train,
    solution=func,
    num_test=nx_test ** 2,
)

net = dde.nn.FNN(
    [2] + [num_dense_nodes] * num_dense_layers + [1], activation, "Glorot uniform"
)

if hard_constraint == True:
    net.apply_output_transform(transform)

model = dde.Model(data, net)

if hard_constraint == True:
    model.compile("adam", lr=learning_rate, metrics=["l2 relative error"])
else:
    loss_weights = [1, weights]
    model.compile(
        "adam",
        lr=learning_rate,
        metrics=["l2 relative error"],
        loss_weights=loss_weights,
    )
    
import horovod.tensorflow as hvd    
resampler = dde.callbacks.PDEResidualResampler(period= epochs // hvd.size())
losshistory, train_state = model.train(epochs=epochs, callbacks = [resampler])
```

### CODE REVIEW

**Comment by pescap** [comment]
#579 #39

**Comment by pescap** [comment]
FYI: https://github.com/horovod/horovod/issues/114

I am currently working on it. The first milestone would be to dispatch the training points over the GPUs. Maybe via a data generator: 

1. generate n_GPU*random_data of same size [N_domain, N_boundary]
2. or just generate n_GPU*random_data of same size [N_domain] and use the same boundary points for all the GPUs

Next, as soon as the horovod architecture is deployed, one could go towards domain decomposition-based solutions.

**Comment by pescap** [comment]
Notes (work in progress): 

Train next batch for PDEs:
https://github.com/lululxvi/deepxde/blob/770e7c1f703682633fe182a6de984987fd579afa/deepxde/data/pde.py#L168

**Comment by pescap** [comment]
I'll start from the beginning, as this PR is too outdated. A quick update:

- I am more familiar with `tf.compat.v1` and it has more models implemented. Therefore, I will implement the Horovod acceleration for this backend first (and not `tensorflow`, as for this original draft)
- At the end, the data point generation is straightforward. One uses random o pseudo-random generation, and each GPU is assigned its own `data`. Uniform distribution would lead to to the same training points over all GPUs...
- I keep accelerating the domain collocation points. I assume that the boundary points do not need to be distributed over GPU.

**Comment by lululxvi** [comment]
@pescap You can also check https://github.com/lululxvi/deepxde/pull/1094, although it is only for backend paddle, but it has many useful functions.

**Comment by pescap** [comment]
> @pescap You can also check #1094, although it is only for backend paddle, but it has many useful functions.

Thank you, I'll try to follow the same structure.

File: deepxde/config.py [hunk 1]
```diff
@@ -7,6 +7,7 @@
 from .backend import backend_name, tf, torch, paddle
 from .real import Real
 
+hvd_dist = False
```
Comment (pescap) [inline deepxde/config.py:10]: @lululxvi I'll keep using this variable to add or not Horovod. Let me know if the variable name is fine.

Comment (lululxvi) [reply deepxde/config.py:10]: We can use this for now.

**Pull request closed without merging**